### DISCUSSION TITLE
"Fast digital pin must be a constant' question constant in class

### DISCUSSION THREAD

**Discussion by yuranikspb** [SpenceKonde/megaTinyCore#445] [Q&A]
```
////////////sketch
#include <lin_stack_nilab.h>

const byte faultTXE_pin=3;
const byte TX_pin=5;
const byte ident = 0x11;

lin_stack_nilab LIN(faultTXE_pin,TX_pin,ident);



//.h

class lin_stack_nilab
{
	public:
		// Constructors
		lin_stack_nilab(const byte tr_p,const byte tx_p, byte ident); 
	
	private:
	const byte tr_pin;// pin transceiver _mode
	const byte tx_pin;// pin transceiver _mode
	
}	
//cpp
// CONSTRUCTORS

lin_stack_nilab::lin_stack_nilab(const byte tr_p,const byte tx_p, byte ident):tr_pin(tr_p),tx_pin(tx_p){	
	identByte = ident; // saving idet to private variable
	//tr_pin=tr_p;
	//tx_pin=tx_p;
	pinMode(tr_pin,OUTPUT);
	transceiver_mode(1); 
	
}


// Method

int lin_stack_nilab::transceiver_mode(byte transceiver_mode_state){
	if(transceiver_mode_state==1)
	{ // Go to Normal mode
		digitalWriteFast(tr_pin,1);
	}else{
	    digitalWriteFast(tr_pin,0);
	}
	delayMicroseconds(20); 
	return 1;
}
```

why not work ? tr_pin  constant.
need compile config to  -O3 ?

**Comment by SpenceKonde** [comment]
The optimizer is not smart enough to perform much of any optimization the moment classes come into play. That's why libraries that involve classes tend to use so much flash, as I understand it. 

digitalWriteFast needs the pin to really be known (no allowance for another instance of a class with a different constant pin - that'd have to have it's own copy of that method: the instruction digitalWriteFast compiles to is a cbi/sbi instruction, and the opcode contains the number of the register to operate on within the low IO space (0-31) and the bit (0-7).... I am not very good with C++ classes so I can't offer much more on whether there's a way to make the optimizer do it.... But that's why it's that way. If you want to flip a pin in a single clock cycle, that's what it takes :-P
